File: scripts/genhelp.py
#!/usr/bin/env python
import sys, os, tempfile, subprocess


def usage(ex):
    [sys.stderr, sys.stdout][not ex].write(
        "Usage: %s srcdir helptexts > helptext.c\n"
            % os.path.basename(sys.argv[0]))
    sys.exit(ex)
if '--help' in sys.argv:
    usage(0)
if len(sys.argv) < 3:
    usage(1)


def die_at(filename, line, message):
    sys.stderr.write("%s:%d: %s\n" % (filename, line + 1, message))
    sys.exit(1)

# valid characters to start a line (see page_help.c)
typechars = frozenset("|+:;!%#=")

srcdir = sys.argv[1]
helptexts = sys.argv[2:]
o = sys.stdout

if sys.version_info[0] > 2:
    unichr = chr
    mapord = lambda s: s
else:
    mapord = lambda s: map(ord, s)

arraynames = []
o.write("extern const unsigned char *help_text[];\n\n")
for idx, textfile in enumerate(helptexts):
    blank = True
    arrname = ("help_text_%s" % str(idx))
    o.write("static const unsigned char %s[] = {\n" % arrname)
    for lnum, line in enumerate(open(os.path.join(srcdir, textfile), 'rb')):
        blank = False
        
        try:
            line = line.decode('utf8').rstrip('\r\n')
        except UnicodeError:
            die_at(textfile, lnum, "malformed Unicode character")

        if not line:
            continue
        elif line.endswith(' '):
            die_at(textfile, lnum, "trailing whitespace")
        elif len(line) > 76:
            die_at(textfile, lnum, "line is longer than 76 characters")
        elif line[0] not in typechars:
            die_at(textfile, lnum, "line-type character %c is not one of %s"
                   % (line[0], ''.join(typechars)))

        line += '\n'
        try:
            line = (line
                .replace(unichr(0x00B6), unichr(0x14)) # paragraph mark
                .replace(unichr(0x00A7), unichr(0x15)) # section mark (why? I don't know)
                .encode('cp437'))
        except UnicodeError:
            die_at(textfile, lnum, "invalid CP437 character")

        o.write(','.join([str(hex(c)) for c in mapord(line)]) + ',\n')
    if blank:
        die_at(textfile, 0, "file is empty")
    # don't forget the NUL termination
    o.write("0};\n\n")
    arraynames.append(arrname)

o.write("const unsigned char* help_text[] = {\n")
for arrname in arraynames:
    o.write("\t%s,\n" % arrname)
o.write("};\n")

File: scripts/genhelp.pl
#!/usr/bin/perl

use 5.8.0;
use strict;
use warnings;
use Encode;
use utf8;
use open ':encoding(utf-8)';

my $usage = sprintf("Usage: %s srcdir helptexts > helptext.c\n", $0);

if ($#ARGV < 1) {
	print STDERR $usage;
	exit 1;
}

foreach my $i (@ARGV) {
	if ($i eq "--help") {
		print STDOUT $usage;
		exit 0;
	}
}

my $srcdir = shift(@ARGV);
my @helptexts = @ARGV;

my $typechars = "|+:;!\%#=";

my @arrnames = ();

sub die_at {
	my ($filename, $line, $message) = @_;

	die(sprintf("%s:%d: %s\n", $filename, $line + 1, $message));
}

sub str_begins_with_spn {
	my ($str1, $str2, @others) = @_;

	foreach my $char (split('', $str2)) {
		if (rindex($str1, $char, 0) == 0) {
			return 1;
		}
	}

	return 0;
}

print "extern const unsigned char *help_text[];\n\n";

for my $i (0 .. $#helptexts) {
	my $file;
	open($file, $srcdir . "/" . $helptexts[$i]) or die(sprintf("could not open file %s", $srcdir . "/" . $helptexts[$i]));

	my $arrname = sprintf("help_text_%d", $i);

	printf("static const unsigned char %s[] = {\n", $arrname);

	my $blank = 1;
	while (my $line = <$file>) {
		$line =~ s/[\r\n]+//; # strip newline chars

		$blank = 0;

		if (length($line) <= 0) {
			# ignore empty lines
			next;
		} elsif (length($line) > 76) {
			die_at($helptexts[$i], $., "line is longer than 76 characters");
		} elsif ($line =~ / $/) {
			die_at($helptexts[$i], $., "trailing whitespace");
		} elsif (!str_begins_with_spn($line, $typechars)) {
			die_at($helptexts[$i], $., sprintf("line-type character %s is not one of %s", substr($line, 0, 1), $typechars))
		}

		foreach my $codepoint (split('', $line)) {
			$codepoint = ord($codepoint);

			if ($codepoint == 0x00B6) {
				$codepoint = 0x14;
			} elsif ($codepoint == 0x00A7) {
				$codepoint = 0x15;
			}

			printf("0x%02x, ", ord(encode("cp437", chr($codepoint))));
		}

		print("0x0A, \n");
	}

    if ($blank) {
        die_at($helptexts[$i], 0, "file is empty");
    }

    print("0x00};\n\n");
    $arrnames[$i] = $arrname;
}

print("const unsigned char* help_text[] = {\n");
foreach my $s (@arrnames) {
	printf("\t%s,\n", $s);
}
print("};\n");
